crates/generated_parser/src/early_error_checker.rs

```diff
@@ -0,0 +1,205 @@
+use crate::context_stack::{
+    BindingKind, BindingsIndex, BreakOrContinueIndex, ContextMetadata, LabelIndex, LabelKind,
+};
+use crate::declaration_kind::DeclarationKind;
+use crate::early_errors::*;
+use crate::error::{ParseError, Result};
+use ast::source_atom_set::{SourceAtomSet, SourceAtomSetIndex};
+use std::cell::RefCell;
+use std::rc::Rc;
+
+pub trait EarlyErrorChecker<'alloc> {
+    fn context_metadata(&mut self) -> &mut ContextMetadata;
+    fn context_metadata_immutable(&self) -> &ContextMetadata;
```

Comment:
nit: When both are available, the Rust idiom is to provide the immutable version by default, and otherwise append `_mut` to the name. (for example [slice::first](https://doc.rust-lang.org/std/primitive.slice.html#method.first) and [slice::first_mut](https://doc.rust-lang.org/std/primitive.slice.html#method.first_mut))

nit: `fn context_stack_mut(&mut self) -> &mut ContextStackData` ?